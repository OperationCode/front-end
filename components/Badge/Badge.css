.Badge {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  margin: 1rem;
}

.Badge > img,
.Badge > svg {
  fill: currentColor;
  height: 100px;
  margin: 1rem 0;
}

.label {
  text-align: center;
  white-space: nowrap;
}
