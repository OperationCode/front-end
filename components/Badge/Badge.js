import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Badge.css';

Badge.propTypes = {
  className: PropTypes.string,

  // TODO: Create custom proptype accepting only `<svg>` or `<img>`
  // TODO: Once above is done, add integration tests regarding proptype
  icon: PropTypes.element.isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  isImageFirst: PropTypes.bool,
};

Badge.defaultProps = {
  className: undefined,
  isImageFirst: true,
};

function Badge({ className, icon, isImageFirst, label }) {
  return (
    <figure className={classNames(styles.Badge, className)}>
      {isImageFirst ? (
        <>
          {icon}
          <figcaption className={styles.label}>{label}</figcaption>
        </>
      ) : (
        <>
          <figcaption className={styles.label}>{label}</figcaption>
          {icon}
        </>
      )}
    </figure>
  );
}

export default Badge;
