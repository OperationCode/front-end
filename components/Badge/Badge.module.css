.Badge {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  margin: 1rem;
}

.Badge > img,
.Badge > svg {
  fill: currentcolor;
  height: 100px;
  margin: 1rem 0;
}

.label {
  text-align: center;
  white-space: nowrap;
}
