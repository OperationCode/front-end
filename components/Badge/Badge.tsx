import classNames from 'classnames';

export type BadgePropsType = {
  /**
   * SVG icon to be used as the badge.
   */
  icon: React.ReactElement;
  /**
   * Optional label that is rendered with the badge.
   */
  label: string | React.ReactElement;
  /**
   * Applies classnames to the base `figure` element for styling.
   */
  className?: string;
  /**
   * Sets whether the label is rendered above, or below, the badge..
   */
  isImageFirst?: boolean;
};

function Badge({ className = undefined, icon, isImageFirst = true, label }: BadgePropsType) {
  return (
    <figure
      className={classNames(
        'flex',
        'flex-col',
        'justify-items-center',
        'place-content-start',
        'm-2',
        '[&>svg]:fill-current',
        '[&>svg]:my-4',
        '[&>img]:mx-0',
        '[&>img]:h-24',
        '[&>svg]:mx-0',
        '[&>svg]:h-24',
        className,
      )}
    >
      {isImageFirst ? (
        <>
          {icon}
          <figcaption className="text-center whitespace-nowrap">{label}</figcaption>
        </>
      ) : (
        <>
          <figcaption className="text-center whitespace-nowrap">{label}</figcaption>
          {icon}
        </>
      )}
    </figure>
  );
}

export default Badge;
