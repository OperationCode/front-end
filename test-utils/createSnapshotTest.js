/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
