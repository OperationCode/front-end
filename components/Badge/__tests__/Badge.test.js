/* eslint-env jest */
import React from 'react';
import { mount } from 'enzyme';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import Icon from 'static/images/icons/github_logo.svg';

import Badge from '../Badge';

const badgeIcon = <Icon />;

describe('Badge', () => {
  it('should render with required props', () => {
    createSnapshotTest(<Badge icon={badgeIcon} label="Testing" />);
  });

  it('should render with many props assigned', () => {
    createSnapshotTest(<Badge icon={badgeIcon} label="Badge Icon" className="test-class" />);
  });

  it('should render the image after the label when `isImageFirst` is true', () => {
    const wrapper = mount(<Badge icon={badgeIcon} label="Badge Icon" isImageFirst />);
    const iconIsFirst = wrapper
      .find('figure')
      .childAt(0)
      .is(Icon);
    expect(iconIsFirst).toBe(true);
  });
  it('should render the image before the label when `isImageFirst` is false', () => {
    const wrapper = mount(<Badge icon={badgeIcon} label="Badge Icon" isImageFirst={false} />);
    const figcaptionIsFirst = wrapper
      .find('figure')
      .childAt(0)
      .is('figcaption');
    expect(figcaptionIsFirst).toBe(true);
  });
});
