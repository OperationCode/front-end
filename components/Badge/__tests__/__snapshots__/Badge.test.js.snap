// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render with many props assigned 1`] = `
<figure
  className="Badge test-class"
>
  <svg />
  <figcaption
    className="label"
  >
    Badge Icon
  </figcaption>
</figure>
`;

exports[`Badge should render with required props 1`] = `
<figure
  className="Badge"
>
  <svg />
  <figcaption
    className="label"
  >
    Testing
  </figcaption>
</figure>
`;
